File: Open-CAF-verification-pipeline copy/backend/app/schemas/assessment.py
"""Pydantic Schemas for CAF Assessments, Outcomes, and Score Summaries."""
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Dict
from uuid import UUID
from datetime import datetime

from app.models.assessment import AssessmentStatus, OutcomeStatus
from app.models.caf import IGPLevel

class AssessmentCreate(BaseModel):
    title: str = Field(..., min_length=3, max_length=255, description="e.g. 2026 Q1 Local Government Assurance")
    scope_description: str = Field(..., min_length=10, description="Systems, networks, and services within scope")
    council_service_name: Optional[str] = Field(None, max_length=255, description="e.g. Adult Social Care Systems")

class AssessmentUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=3, max_length=255)
    scope_description: Optional[str] = None
    status: Optional[AssessmentStatus] = None
    council_service_name: Optional[str] = None

class OutcomeEvaluationUpdate(BaseModel):
    status: OutcomeStatus
    assessor_rationale: Optional[str] = None
    reviewer_notes: Optional[str] = None
    satisfied_igp_ids: Optional[List[UUID]] = None

class IGPCheckResponse(BaseModel):
    igp_id: UUID
    level: IGPLevel
    description: str
    sort_order: int
    is_satisfied: bool

    model_config = ConfigDict(from_attributes=True)

class AssessmentOutcomeResponse(BaseModel):
    id: UUID
    assessment_id: UUID
    outcome_id: str
    outcome_title: Optional[str] = None
    outcome_description: Optional[str] = None
    principle_id: Optional[str] = None
    objective_id: Optional[str] = None
    status: OutcomeStatus
    assessor_rationale: Optional[str] = None
    reviewer_notes: Optional[str] = None
    assessed_at: Optional[datetime] = None
    igp_checks: List[IGPCheckResponse] = []

    model_config = ConfigDict(from_attributes=True)

class PrincipleScore(BaseModel):
    principle_id: str
    principle_title: str
    total_outcomes: int
    achieved: int
    partially_achieved: int
    not_achieved: int
    not_started: int
    completion_rate: float

class ObjectiveScore(BaseModel):
    objective_id: str
    objective_title: str
    total_outcomes: int
    achieved: int
    partially_achieved: int
    not_achieved: int
    not_started: int
    completion_rate: float
    principles: Dict[str, PrincipleScore] = {}

class ScoreSummary(BaseModel):
    total_outcomes: int = 39
    evaluated_outcomes: int
    remaining_outcomes: int
    completion_rate: float
    overall_achieved_count: int
    overall_achieved_pct: float
    overall_partially_achieved_count: int
    overall_partially_achieved_pct: float
    overall_not_achieved_count: int
    overall_not_achieved_pct: float
    overall_not_started_count: int
    objectives: Dict[str, ObjectiveScore] = {}

class AssessmentResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    title: str
    scope_description: str
    status: AssessmentStatus
    council_service_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    score_summary: Optional[ScoreSummary] = None

    model_config = ConfigDict(from_attributes=True)

class AssessmentDetailResponse(AssessmentResponse):
    outcomes: List[AssessmentOutcomeResponse] = []

File: Open-CAF-verification-pipeline copy/frontend/src/types/assessment.ts
/**
 * Frontend Type Definitions for CAF Assessments, Outcomes, and Scoring
 */

export type OutcomeStatus = 'NOT_STARTED' | 'ACHIEVED' | 'PARTIALLY_ACHIEVED' | 'NOT_ACHIEVED';
export type IGPLevel = 'ACHIEVED' | 'PARTIALLY_ACHIEVED';
export type AssessmentFilter = 'all' | 'unassessed' | 'needs_evidence' | 'deficits_only';

export interface IGPCheck {
  igp_id: string;
  level: IGPLevel;
  description: string;
  sort_order: number;
  is_satisfied: boolean;
}

export interface AssessmentOutcome {
  id: string;
  assessment_id: string;
  outcome_id: string;
  outcome_title?: string;
  outcome_description?: string;
  principle_id?: string;
  objective_id?: string;
  status: OutcomeStatus;
  assessor_rationale?: string;
  reviewer_notes?: string;
  assessed_at?: string;
  igp_checks: IGPCheck[];
  evidence_count?: number;
  evidence_files?: Array<{ id: string; filename: string; category?: string }>;
}

export interface PrincipleScore {
  principle_id: string;
  principle_title: string;
  total_outcomes: number;
  achieved: number;
  partially_achieved: number;
  not_achieved: number;
  not_started: number;
  completion_rate: number;
}

export interface ObjectiveScore {
  objective_id: string;
  objective_title: string;
  total_outcomes: number;
  achieved: number;
  partially_achieved: number;
  not_achieved: number;
  not_started: number;
  completion_rate: number;
  principles: Record<string, PrincipleScore>;
}

export interface ScoreSummary {
  total_outcomes: number;
  evaluated_outcomes: number;
  remaining_outcomes: number;
  completion_rate: number;
  overall_achieved_count: number;
  overall_achieved_pct: number;
  overall_partially_achieved_count: number;
  overall_partially_achieved_pct: number;
  overall_not_achieved_count: number;
  overall_not_achieved_pct: number;
  overall_not_started_count: number;
  overall_maturity_score?: number;
  objectives: Record<string, ObjectiveScore>;
}

export interface AssessmentResponse {
  id: string;
  tenant_id: string;
  title: string;
  scope_description: string;
  status: 'DRAFT' | 'IN_PROGRESS' | 'UNDER_REVIEW' | 'COMPLETED' | 'ARCHIVED';
  council_service_name?: string;
  created_at: string;
  updated_at: string;
  score_summary?: ScoreSummary;
}

export interface AssessmentDetail extends AssessmentResponse {
  outcomes: AssessmentOutcome[];
}
